Add official .NET obsoletion policy (#31104)
dotnet/docs: docs/core/compatibility/index.md

- ❌ **DISALLOWED: Renaming or removing a public type**

  This breaks all code that uses the renamed or removed type.

>   .NET's policy for obsoleting APIS is as follows:

Reviewer: ```suggestion
  .NET's policy for obsoleting APIs is as follows:
```

>   > An API that shipped in a long-term support (LTS) release must be obsoleted in the subsequent LTS release before it can be removed. In rare cases, exceptions are made to obsolete an API before the subsequent LTS release based on business needs. These obsoletions will be documented and communicated to customers.

Reviewer 2: @gewarren Can we reword the last statement to "All obsoletions are documented and communicated to customers."?  I had meant to say that exceptions will not be a surprise to customers, but in the current form it might be interpreted to mean that we will only document obsoletions that are not in LTS release.

>   > An API that shipped in a long-term support (LTS) release must be obsoleted in the subsequent LTS release before it can be removed. In rare cases, exceptions are made to obsolete an API before the subsequent LTS release based on business needs. These obsoletions will be documented and communicated to customers.

Author: ```suggestion
  > An API that shipped in a long-term support (LTS) release must be obsoleted in the subsequent LTS release before it can be removed. In rare cases, exceptions are made to obsolete an API before the subsequent LTS release based on business needs. All obsoletions are documented and communicated to customers.
```
